enhancements/network/routed-ingress-support-for-primary-udn-attached-vms-with-static-ips.md

```diff
@@ -0,0 +1,789 @@
+particular annotation on the pod.
+
+The flow is described in more detail in the
+[De-centralized IP management workflow](#de-centralized-ip-management-workflow) section.
+
+##### De-centralized IP management workflow
+
+The [de-centralized IP management](#de-centralized-ip-management) flow is
+described in the following sequence diagram:
+
+```mermaid
+sequenceDiagram
+actor Admin
+actor VM Owner
+
+participant MTV
+participant CNV
+participant o as OVN-Kubernetes
+
+Admin ->> o: cluster UDN
+o -->> Admin: OK
+
+Admin ->> CNV: provision IPAMClaim w/ MAC and IP requests
+CNV -->> Admin: OK
+
+VM Owner ->> MTV: import VM
+MTV ->> CNV: create VM(name=<...>, primaryUDNMac=origMAC)
+CNV ->> CNV: ips = getIPsForMAC(mac=origMAC)
+CNV ->> o: create pod(mac=origMAC, IPs=ips)
+o -->> CNV: OK
+CNV -->> MTV: OK
+MTV -->> VM Owner: OK
+```
+
+Hence, the required changes would be:
+- the `IPAMClaim` CRD status sub-resource will need to be updated, adding a
+  `Conditions` array. This way we will be able to report errors back to the
+  user (e.g. the desired IP allocation is already in use).
+- the `IPAMClaim` CRD spec will need to be updated, adding `MAC` and `IPs`
+  requests.
+- OVN-Kubernetes will need to try to apply the MAC and IP addresses requested
+  in its spec. If any conflicts are found for either MAC or IPs, the error will
+  be persisted in the `IPAMClaim` status (in the conditions array).
+- OVN-Kubernetes will record the requested IPs in the status sub-resource
+  (already does so today), along with a `Success` condition.
+
+This option has the network admin create the IPAMClaim, and request a MAC and
+IP addresses for the VM; in a second stage of the implementation, MTV (or any
+other component able to introspect the source cluster for the VM's MAC and IPs)
+will create the IPAMClaim on behalf of the admin user.
+
+###### Admin user flows for de-centralized approach
+
+The admin user will have to perform the following operations:
+- provision the cluster UDN CR having the overlay definition.
+  - layer2 topology ; IPAM lifecycle `Persistent` are a **must**
+- provision the `IPAMClaim` object, whose name **must** be created like
+  `<vm name>.<logical network name>`. The `IPAMClaim` must have both the IP and
+  MAC addresses for the VM attachment. This is only required until another tool
+  automatically provisions the `IPAMClaim`s on behalf of the admin
+
+As indicated above, when a tool exists that can introspect the source cluster
+to learn the VM's IP addresses, the admin user will only require provisioning
+the cluster UDN CR.
+
+###### VM owner flows for de-centralized approach
+
+The VM owner just has to use MTV to import the VM into CNV.
+
+## Test Plan
+
+* E2E upstream and downstream jobs covering VM creation with requested
+  IP/MAC/gateway configuration for both IPv4 and dual-stack configurations
+* E2E downstream jobs covering a VM (having a network configuration which must
```

Comment:
@mnecas is MTV a part of any OCP CI? I am wondering if an automated E2E in OCP CI is realistic considering that MTV might require an existing VMWare deployment.

Reply:
sadly it's not, know there is some effort to add it to kubevirt CI but nothing as of the moment

Reply:
I think it is crucial to require a simulation in which MTV imports some data and exports it in a form that OCP-V can consume. The simulation can be as simple as pre-prepared data files with examples of other VMM systems.

If we will cover tests only at  project borders, we miss the actual end-to-end understanding.
I would say that an e2e project level approach is fine for U/S, but it is not enough for D/S. On D/S you could run those U/S tests as tier-1 and compose solution level testing as tier-2. I.e. tier-2 should include the user perspective of triggering MTV with specific data as input and checking the expected end result (VMs running) is achieved. As mentioned before, I think we need tooling to allow simulated platforms (e.g. to avoid the need to have a specific vendor VMM).

Reply:
+1 it is very critical to have full blown integration test for all pieces coming together